frame/safe-mode/src/lib.rs

```diff
@@ -0,0 +1,522 @@
+		/// block specified.
+		///
+		/// ### Safety
+		///
+		/// Can only be called by the [`Config::ForceReservationOrigin`] origin.
+		#[pallet::weight(T::WeightInfo::force_release_reservation())]
+		pub fn force_release_reservation(
+			origin: OriginFor<T>,
+			account: T::AccountId,
+			block: T::BlockNumber,
+		) -> DispatchResult {
+			T::ForceReservationOrigin::ensure_origin(origin)?;
+
+			Self::do_release(None, account, block)
+		}
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		/// Automatically deactivates the safe-mode when the period runs out.
+		///
+		/// Bypasses any call filters to avoid getting rejected by them.
+		fn on_initialize(current: T::BlockNumber) -> Weight {
+			match ActiveUntil::<T>::get() {
+				Some(limit) if current > limit => {
+					let _ = Self::do_deactivate(ExitReason::Timeout)
+						.defensive_proof("Must be inactive; qed");
+					T::DbWeight::get().reads_writes(1, 1)
+				},
+				_ => T::DbWeight::get().reads(1), // TODO benchmark
+			}
+		}
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	/// Logic for the [`crate::Pallet::activate`] and [`crate::Pallet::force_activate`] calls.
+	fn do_activate(who: Option<T::AccountId>, duration: T::BlockNumber) -> DispatchResult {
+		if let Some(who) = who {
+			let reserve = T::ActivateReservationAmount::get().ok_or(Error::<T>::NotConfigured)?;
+			Self::reserve(who, reserve)?;
+		}
+
+		ensure!(!ActiveUntil::<T>::exists(), Error::<T>::IsActive);
+		let limit = <frame_system::Pallet<T>>::block_number().saturating_add(duration);
+		ActiveUntil::<T>::put(limit);
+		Self::deposit_event(Event::Activated { block: limit });
+		Ok(())
+	}
+
+	/// Logic for the [`crate::Pallet::extend`] and [`crate::Pallet::force_extend`] calls.
+	fn do_extend(who: Option<T::AccountId>, duration: T::BlockNumber) -> DispatchResult {
+		if let Some(who) = who {
+			let reserve = T::ExtendReservationAmount::get().ok_or(Error::<T>::NotConfigured)?;
+			Self::reserve(who, reserve)?;
+		}
+
+		let limit =
+			ActiveUntil::<T>::take().ok_or(Error::<T>::IsInactive)?.saturating_add(duration);
+		ActiveUntil::<T>::put(limit);
+		Self::deposit_event(Event::<T>::Extended { block: limit });
+		Ok(())
+	}
+
+	/// Logic for the [`crate::Pallet::force_deactivate`] call.
+	///
+	/// Errors if the safe-mode is already inactive.
+	/// Does not check the origin.
+	fn do_deactivate(reason: ExitReason) -> DispatchResult {
+		let _limit = ActiveUntil::<T>::take().ok_or(Error::<T>::IsInactive)?;
+		Self::deposit_event(Event::Exited { reason });
+		Ok(())
+	}
+
+	/// Logic for the [`crate::Pallet::release_reservation`] and
+	/// [`crate::Pallet::force_release_reservation`] calls.
+	///
+	/// Errors if the safe-mode is active with [`Error::IsActive`].
+	/// Errors if release is called too soon by anyone but [`Config::ForceReservationOrigin`] with
+	/// [`Error::CannotReleaseYet`]. Does not check the origin.
+	fn do_release(
+		who: Option<T::AccountId>,
```

Comment:
```suggestion
		force: bool,
```
You dont use the fact that it is an `AccountId`, only whether it is *some*.